farm/data_handler/processor.py

```diff
@@ -1727,6 +1728,292 @@ def _sample_to_features(self, sample) -> dict:
+        positive_context = positive_context[:self.num_positives]
+        hard_negative_context = list(filter(lambda x: x["label"] == "hard_negative", dictionary["passages"]))
+        if self.shuffle_negatives:
+            random.shuffle(hard_negative_context)
+        hard_negative_context = hard_negative_context[:self.num_hard_negatives]
+
+        positive_ctx_titles = [passage.get("title", None) for passage in positive_context]
+        positive_ctx_texts = [passage["text"] for passage in positive_context]
+        hard_negative_ctx_titles = [passage.get("title", None) for passage in hard_negative_context]
+        hard_negative_ctx_texts = [passage["text"] for passage in hard_negative_context]
+
+        # all context passages and labels: 1 for positive context and 0 for hard-negative context
+        ctx_label = [1]*self.num_positives + [0]*self.num_hard_negatives #(self.num_positives if self.num_positives < len(positive_context) else len(positive_context)) + \
+        # +(self.num_hard_negatives if self.num_hard_negatives < len(hard_negative_context) else len(hard_negative_context))
+
+        # featurize the query
+        query_inputs = self.query_tokenizer.encode_plus(
+            text=query,
+            max_length=self.max_seq_len,
+            add_special_tokens=True,
+            truncation_strategy='do_not_truncate',
+            padding="max_length",
+            return_token_type_ids=True,
+        )
+
+        # featurize context passages
+        if self.embed_title:
+            # embed title with positive context passages + negative context passages
+            all_ctx = [tuple((title, ctx)) for title, ctx in
+                       zip(positive_ctx_titles, positive_ctx_texts)] + \
+                      [tuple((title, ctx)) for title, ctx in
+                       zip(hard_negative_ctx_titles, hard_negative_ctx_texts)]
+        else:
+            all_ctx = positive_ctx_texts + hard_negative_ctx_texts
+
+        # assign empty string tuples if hard_negative passages less than num_hard_negatives
+        all_ctx += [('', '')] * ((self.num_positives + self.num_hard_negatives)-len(all_ctx))
+
+        ctx_inputs = self.passage_tokenizer.batch_encode_plus(
+            all_ctx,
+            add_special_tokens=True,
+            truncation=True,
+            padding="max_length",
+            max_length=self.max_seq_len,
+            return_token_type_ids=True
+        )
+
+        query_input_ids, query_segment_ids, query_padding_mask = query_inputs["input_ids"], query_inputs[
+                                                            "token_type_ids"], query_inputs["attention_mask"]
+        ctx_input_ids, ctx_segment_ids_, ctx_padding_mask = ctx_inputs["input_ids"], ctx_inputs["token_type_ids"], \
+                                                           ctx_inputs["attention_mask"]
+        ctx_segment_ids = list(torch.zeros((len(ctx_segment_ids_), len(ctx_segment_ids_[0]))).numpy())
```

Comment:
I believe this was one of the design choices we discussed. Can you also add a quick note here stating that all-zero segment ids is similar to the original DPR implementation but might be worth an experiment later on